File: flutter_app/lib/services/profile_service.dart
import '../config/api_config.dart';
import '../models/user_profile.dart';
import 'api_client.dart';
import 'base_api_service.dart';

class ProfileService extends BaseApiService {
  final ApiClient apiClient;

  ProfileService(this.apiClient) : super(apiClient);

  Future<UserProfile> fetchProfile() async {
    try {
      final response = await apiClient.get(
        ApiConfig.profileMeEndpoint,
        context: 'ProfileService.fetchProfile',
      );
      if (response is Map<String, dynamic>) {
        return UserProfile.fromJson(response);
      }
      throw Exception('Unexpected profile response: $response');
    } catch (e, st) {
      handleError('Failed to fetch profile', e, st);
      rethrow;
    }
  }

  Future<UserProfile> fetchProfileById(String userId) async {
    try {
      final response = await apiClient.get(
        ApiConfig.profileByIdEndpoint(userId),
        context: 'ProfileService.fetchProfileById',
      );
      if (response is Map<String, dynamic>) {
        return UserProfile.fromJson(response);
      }
      throw Exception('Unexpected profile response: $response');
    } catch (e, st) {
      handleError('Failed to fetch profile by id', e, st);
      rethrow;
    }
  }

  Future<UserProfile> updateCoachingProfile({
    bool? enabled,
    bool? acceptingClients,
    String? tagline,
    String? description,
    List<String>? specializations,
    List<String>? languages,
    int? experienceYears,
    String? timezone,
    CoachingRatePlan? ratePlan,
  }) async {
    final payload = <String, dynamic>{};
    if (enabled != null) payload['enabled'] = enabled;
    if (acceptingClients != null) payload['accepting_clients'] = acceptingClients;
    if (tagline != null) payload['tagline'] = tagline;
    if (description != null) payload['description'] = description;
    if (specializations != null) payload['specializations'] = specializations;
    if (languages != null) payload['languages'] = languages;
    if (experienceYears != null) payload['experience_years'] = experienceYears;
    if (timezone != null) payload['timezone'] = timezone;
    if (ratePlan != null) payload['rate_plan'] = ratePlan.toJson();

    try {
      final response = await apiClient.patch(
        ApiConfig.profileMeCoachingEndpoint,
        payload,
        context: 'ProfileService.updateCoachingProfile',
      );
      if (response is Map<String, dynamic>) {
        return UserProfile.fromJson(response);
      }
      throw Exception('Unexpected coaching profile update response: $response');
    } catch (e, st) {
      handleError('Failed to update coaching profile', e, st);
      rethrow;
    }
  }

  Future<UserProfile> updateProfile({
    String? displayName,
    String? bio,
    bool? isPublic,
    double? bodyweightKg,
    double? heightCm,
    int? age,
    String? sex,
    double? trainingExperienceYears,
    String? trainingExperienceLevel,
    String? primaryDefaultGoal,
    String? trainingEnvironment,
  }) async {
    final payload = <String, dynamic>{};
    if (displayName != null) payload['display_name'] = displayName;
    if (bio != null) payload['bio'] = bio;
    if (isPublic != null) payload['is_public'] = isPublic;
    if (bodyweightKg != null) payload['bodyweight_kg'] = bodyweightKg;
    if (heightCm != null) payload['height_cm'] = heightCm;
    if (age != null) payload['age'] = age;
    if (sex != null) payload['sex'] = sex;
    if (trainingExperienceYears != null) {
      payload['training_experience_years'] = trainingExperienceYears;
    }
    if (trainingExperienceLevel != null) {
      payload['training_experience_level'] = trainingExperienceLevel;
    }
    if (primaryDefaultGoal != null) {
      payload['primary_default_goal'] = primaryDefaultGoal;
    }
    if (trainingEnvironment != null) {
      payload['training_environment'] = trainingEnvironment;
    }

    try {
      final response = await apiClient.patch(
        ApiConfig.profileMeEndpoint,
        payload,
        context: 'ProfileService.updateProfile',
      );
      if (response is Map<String, dynamic>) {
        return UserProfile.fromJson(response);
      }
      throw Exception('Unexpected profile update response: $response');
    } catch (e, st) {
      handleError('Failed to update profile', e, st);
      rethrow;
    }
  }

  Future<UserSettings> updateSettings({
    String? unitSystem,
    String? locale,
    String? timezone,
    bool? notificationsEnabled,
  }) async {
    final payload = <String, dynamic>{};
    if (unitSystem != null) payload['unit_system'] = unitSystem;
    if (locale != null) payload['locale'] = locale;
    if (timezone != null) payload['timezone'] = timezone;
    if (notificationsEnabled != null) payload['notifications_enabled'] = notificationsEnabled;

    try {
      final response = await apiClient.patch(
        ApiConfig.profileSettingsEndpoint,
        payload,
        context: 'ProfileService.updateSettings',
      );
      if (response is Map<String, dynamic>) {
        return UserSettings.fromJson(response);
      }
      throw Exception('Unexpected settings update response: $response');
    } catch (e, st) {
      handleError('Failed to update settings', e, st);
      rethrow;
    }
  }
}

File: services/accounts-service/accounts_service/services/profile_service.py
from __future__ import annotations

from dataclasses import dataclass

from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from ..models import UnitSystem, UserCoachingProfile, UserProfile, UserSettings
from ..schemas import (
    CoachingProfileResponse,
    CoachingRatePlan,
    ProfileResponse,
    SettingsResponse,
)


@dataclass
class ProfileData:
    profile: UserProfile
    settings: UserSettings
    coaching: UserCoachingProfile | None = None


async def _fetch_profile(db: AsyncSession, user_id: str) -> UserProfile | None:
    return await db.get(UserProfile, user_id)


async def _fetch_settings(db: AsyncSession, user_id: str) -> UserSettings | None:
    return await db.get(UserSettings, user_id)


async def _fetch_coaching_profile(db: AsyncSession, user_id: str) -> UserCoachingProfile | None:
    return await db.get(UserCoachingProfile, user_id)


async def ensure_profile_and_settings(db: AsyncSession, user_id: str) -> ProfileData:
    profile = await _fetch_profile(db, user_id)
    settings = await _fetch_settings(db, user_id)
    created = False
    if profile is None:
        profile = UserProfile(user_id=user_id, display_name=user_id)
        db.add(profile)
        created = True
    if settings is None:
        settings = UserSettings(user_id=user_id)
        db.add(settings)
        created = True
    if created:
        await db.commit()
        await db.refresh(profile)
        await db.refresh(settings)
    coaching = await _fetch_coaching_profile(db, user_id)
    return ProfileData(profile=profile, settings=settings, coaching=coaching)


async def ensure_coaching_profile(db: AsyncSession, user_id: str) -> UserCoachingProfile:
    coaching = await _fetch_coaching_profile(db, user_id)
    if coaching is None:
        coaching = UserCoachingProfile(user_id=user_id)
        db.add(coaching)
        await db.flush()
    return coaching


def build_profile_response(data: ProfileData) -> ProfileResponse:
    return ProfileResponse(
        user_id=data.profile.user_id,
        display_name=data.profile.display_name,
        bio=data.profile.bio,
        photo_url=data.profile.photo_url,
        bodyweight_kg=data.profile.bodyweight_kg,
        height_cm=data.profile.height_cm,
        age=data.profile.age,
        sex=data.profile.sex,
        training_experience_years=data.profile.training_experience_years,
        training_experience_level=data.profile.training_experience_level,
        primary_default_goal=data.profile.primary_default_goal,
        training_environment=data.profile.training_environment,
        weekly_gain_coef=data.profile.weekly_gain_coef,
        last_active_at=data.profile.last_active_at,
        is_public=data.profile.is_public,
        created_at=data.profile.created_at,
        updated_at=data.profile.updated_at,
        settings=SettingsResponse(
            unit_system=data.settings.unit_system.value,
            locale=data.settings.locale,
            timezone=data.settings.timezone,
            notifications_enabled=data.settings.notifications_enabled,
            created_at=data.settings.created_at,
            updated_at=data.settings.updated_at,
        ),
        coaching=build_coaching_response(data.coaching),
    )


def build_coaching_response(
    entity: UserCoachingProfile | None,
) -> CoachingProfileResponse | None:
    if entity is None:
        return None

    rate_plan = None
    if any([entity.rate_type, entity.rate_currency, entity.rate_amount_minor is not None]):
        rate_plan = CoachingRatePlan(
            type=entity.rate_type,
            currency=entity.rate_currency,
            amount_minor=entity.rate_amount_minor,
        )

    return CoachingProfileResponse(
        enabled=bool(entity.enabled),
        accepting_clients=bool(entity.accepting_clients),
        tagline=entity.tagline,
        description=entity.description,
        specializations=list(entity.specializations or []),
        languages=list(entity.languages or []),
        experience_years=entity.experience_years,
        timezone=entity.timezone,
        rate_plan=rate_plan,
        created_at=entity.created_at,
        updated_at=entity.updated_at,
    )


def parse_unit_system(raw: str | None) -> UnitSystem | None:
    if raw is None:
        return None
    try:
        return UnitSystem(raw)
    except ValueError as exc:  # pragma: no cover - validation error path
        allowed = ", ".join(u.value for u in UnitSystem)
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"unit_system must be one of: {allowed}",
        ) from exc
